### PR TITLE
Dev/0.1.0

### PR DESCRIPTION
**Pull request by adatzer**

### CODE REVIEW

File: CONTRIBUTING.md
Comment (jethron) [inline CONTRIBUTING.md:71]: Nit: Is it still a dedicated _wiki_ page? What do we even call the docs site pages?

File: .github/workflows/cd.yml
Comment (jethron) [inline .github/workflows/cd.yml:40]: Not sure what this protects against, do we need to lint the sections of `template.tpl` as well?

Looks like we can split it pretty easily with:
```shell
csplit template.tpl --suppress-matched 1 '/^___\([A-Z_]\+\)___$/' '{*}'
```

Which gives us:
```shell
file xx*
xx00: empty
xx01: ASCII text
xx02: JSON text data
xx03: JSON text data
xx04: JavaScript source, ASCII text
xx05: JSON text data
xx06: ASCII text, with very long lines (1350)
xx07: ASCII text
```
(where xx06 is also YAML).

Output doesn't seem very useful though, so maybe not worth it.

```shell
python3 -m yamllint -d relaxed xx02 xx03 xx05 xx06
xx02
  1:1       warning  too many blank lines (1 > 0)  (empty-lines)
  13:81     warning  line too long (150 > 80 characters)  (line-length)
  19:1      warning  too many blank lines (2 > 0)  (empty-lines)

xx03
  1:1       warning  too many blank lines (1 > 0)  (empty-lines)
  13:81     warning  line too long (129 > 80 characters)  (line-length)
  32:81     warning  line too long (153 > 80 characters)  (line-length)
  51:81     warning  line too long (87 > 80 characters)  (line-length)
  58:81     warning  line too long (93 > 80 characters)  (line-length)
  78:81     warning  line too long (331 > 80 characters)  (line-length)
  118:81    warning  line too long (111 > 80 characters)  (line-length)
  126:81    warning  line too long (399 > 80 characters)  (line-length)
  166:81    warning  line too long (153 > 80 characters)  (line-length)
  173:81    warning  line too long (116 > 80 characters)  (line-length)
  210:81    warning  line too long (145 > 80 characters)  (line-length)
  215:81    warning  line too long (140 > 80 characters)  (line-length)
  225:1     warning  too many blank lines (2 > 0)  (empty-lines)

xx05
  1:1       warning  too many blank lines (1 > 0)  (empty-lines)
  162:1     warning  too many blank lines (2 > 0)  (empty-lines)

xx06
  1:1       warning  too many blank lines (1 > 0)  (empty-lines)
  383:81    warning  line too long (82 > 80 characters)  (line-length)
  833:81    warning  line too long (113 > 80 characters)  (line-length)
  871:1     warning  too many blank lines (2 > 0)  (empty-lines)
```
But this could let us run lints on `xx04` which is the actual JS code which might be handy; could in theory even do some type checking or something in the long term.

File: template.tpl
Comment (jethron) [inline template.tpl:659]: Should the fallback here be using `getTimestamp()` instead of `getTimestampMillis()` so it's more stable? The API says this is used for retries so making it use current time will change the ID each time right?

File: template.tpl
Comment (jethron) [inline template.tpl]: So this should be like the tag's version? Should it be more prominent/near the top of the code then?

File: template.tpl
Comment (jethron) [inline template.tpl]: Docs say `metricValue` is required for metrics but LD won't validate the payload, should _we_ be checking it's a number and erroring if not? Or does `makeNumber()` already fail like this?

File: template.tpl
Comment (jethron) [inline template.tpl:744]: The [docs](https://developers.google.com/tag-platform/tag-manager/server-side/api#sendhttprequest) don't seem to match this signature anymore, should we switch to the new promise-based version at some point?

File: template.tpl
Comment (jethron) [inline template.tpl:699]: Have they given us any detail on what `anonymousUser` is supposed to be or are we just working from the public docs too?

**Review by jethron — COMMENTED** [review]
Changes seem good, but I have some more general questions as well.

**Comment by jethron** [comment]
Oh, also the readme should probably link explicitly to https://docs.launchdarkly.com/home/creating-experiments/import-metric-events somewhere.

File: CONTRIBUTING.md
Comment (adatzer) [reply CONTRIBUTING.md:71]: You are right, there is only docs now, so not sure which would be the right term for it. This page is common in many of our repositories, so may be a good opportunity to find out if we could improve it.

File: .github/workflows/cd.yml
Comment (adatzer) [reply .github/workflows/cd.yml:40]: The reason we lint the `metadata.yaml` is because an unnoticed syntax error may result in the template vanishing from the gallery. On the template file, what you described above is a great idea! As you said it may be handy for more things so definitely worth considering going forwards! At the moment, since the template is being developed in the template editor and exported it sounds safe to assume it is valid at least from lint perspective?

File: template.tpl
Comment (adatzer) [reply template.tpl]: This version is part of the `User-Agent` header, about which the docs only say `VERSION can be any format, but you should update it if you make major changes to your implementation.`
If i understand this correctly, it is probably not correlated to our tag version, but could potentially relate to tag configuration (e.g. a change to the value a metric corresponds, that could justify a "major" change). I am not sure about it though, it feels a bit vague. For example, could we just hardcode it to `1` and never change it? Would it ever be a request to make it configurable? Do you think it is something we could leave as is for `0.1.0` release?

File: template.tpl
Comment (adatzer) [reply template.tpl]: `makeNumber` will not fail. What it returns is always of type `number` (per `getType`) but it can "mean" `NaN`.
It is indeed an option to fail here. Would we want this to be a concern of the tag?

File: template.tpl
Comment (adatzer) [reply template.tpl:699]: We just work on the public docs for now.

File: template.tpl
Comment (adatzer) [reply template.tpl:659]: Exactly, if there is no `x-sp-event_id` then yes. The tag will not retry (at most it will log the response). Do you think it is ok to keep using this as generic payload id for non Snowplow events?

Comment (jethron) [reply template.tpl:659]: Yeah if there's no retries I think this is fine, just semantics I don't think need to actually change.

File: template.tpl
Comment (jethron) [reply template.tpl]: Maybe `1-{{Container Version}}` to capture potential reconfiguration? I doubt LD do much with this anyway so it's probably not that important to get right. I agree `1` is OK for this early release. :)

File: template.tpl
Comment (jethron) [reply template.tpl]: `NaN` will `JSON.stringify()` to `null` though. I think maybe we should fail here since the only alternative is silently failing on the LD side; better for someone to see the errors and have a chance to specify a default in their configuration or something, I think.

File: template.tpl
Comment (jethron) [reply template.tpl:699]: Probably one for 0.2.0 I guess, the docs are not super clear on what this should be. I'm thinking if the User ID is still undefined at this point we'd have another field for `data.anonymousUserValueCustom` and use that to populate `anonymousUser` instead? E.g. pull in the Domain User ID or some equivalent.

([docs](https://docs.launchdarkly.com/home/creating-experiments/import-metric-events#request-body-format)):
>A JSON object with two properties named user and anonymousUser. The value of the user property is the context key and must uniquely identify the context that the metric is about. The context key must exactly match the context key provided to any LaunchDarkly SDKs evaluating flags that will be used in the experiments.

File: .github/workflows/cd.yml
Comment (jethron) [reply .github/workflows/cd.yml:40]: Yeah for sure, definitely not something that needs to hold up this PR! :)

I think we can safely assume the contents will always be valid but some future tooling might help us catch some issues GTM's UI doesn't catch in the future as you say.

**Review by igneel64 — APPROVED** [review]

File: template.tpl
Comment (adatzer) [reply template.tpl:744]: This is definitely something to consider for all our tags!

File: template.tpl
Comment (adatzer) [reply template.tpl:699]: Just opened [this issue](https://github.com/snowplow/snowplow-gtm-server-side-launchdarkly-tag/issues/3) for us to investigate for a next release!

**Comment by adatzer** [comment]
Thank you both for reviewing!
Changes since you last viewed:

1. Add link to LaunchDarkly metric import docs
2. Fail on non-number metric value: @igneel64 , i changed my mind on this one since our discussion yesterday! Even though i initially considered to be more liberal into what we send (given similar strategy in other tags as well), i now think @jethron 's point about failing is also valid, especially considering related debugging difficulties we have already faced e.g. in Amplitude tag, for not failing early. So i went for failing. We can definitely reconsider in a next release. What do you think?
3. Add option to specify version: @jethron what do you think about this? We discussed with @igneel64 that it makes sense to allow versioning to be configurable so that people choose what makes sense for them. We also added a value hint towards Container Version variable, which could also be of debugging help.

Steps going forwards, if approved (please let me know if you think otherwise):

- merge `dev/0.1.0` to `release/0.1.0`
- rebase `release/0.1.0` to a single commit and prepare for release
- release
- make repository public
- submit to gtm gallery

**Comment by igneel64** [comment]
> 2. Fail on non-number metric value: @igneel64 , i changed my mind on this one since our discussion yesterday! Even though i initially considered to be more liberal into what we send (given similar strategy in other tags as well), i now think @jethron 's point about failing is also valid, especially considering related debugging difficulties we have already faced e.g. in Amplitude tag, for not failing early. So i went for failing. We can definitely reconsider in a next release. What do you think?

@adatzer 
Does `fail` just log into the preview mode ? Seems harmless to me to add it 👍

**Review by igneel64 — APPROVED** [review]

**Comment by adatzer** [comment]
@igneel64 yes, it fails the tag and logs with the message if logging is enabled (by default in preview).

File: template.tpl
Comment (jethron) [reply template.tpl]: Last comment, promise. :)

I feel like we should still be checking the result of `makeNumber` rather than the input; if it's a string or something it should work unless it's nonsense, right?

Maybe like:
```javascript
  const metricValue = makeNumber(getEventData(data.metricValueCustom));

  if (JSON.stringify(metricValue) === "null") { // check for undefined, NaN, Infinity, etc
    return fail(loggingEnabled, stdLogInfo, {
      msg: 'Metric Value must be a number.',
    });
  }
  ldEvent.metricValue = metricValue;
```

**Review by jethron — APPROVED** [review]
One last comment above on the metric value change, but other changes LGTM.

File: template.tpl
Comment (adatzer) [reply template.tpl]: You are right! If user provides a string that can be parsed into a number, then yes, indeed, we should be able to use it.
There is a lot of pain with `makeNumber`, especially if there are no guarantees on the input.. `json.stringify` is a great idea, however there are some more edge cases. For example, if the input is `null` or empty string or even empty array, then we get zero :sob: .

I went for introducing a new function `safeMakeNumber` (see last commit) as a workaround that covers for all edge cases and at least allows us to be explicit about it. (we could also use it inside `getTimestamp`, if we go for it) 

Let me know what you think!

File: template.tpl
Comment (jethron) [inline template.tpl:616]: Still fails on the `Infinity` and `"Infinity"` cases (produces output will `null` value), but it's super edge case so I don't think it matters.

**Review by jethron — APPROVED** [review]
LGTM!

File: template.tpl
Comment (adatzer) [reply template.tpl:616]: This is true. At the moment the custom value for metric is derived from an event property, so i assume that Infinity cannot be as a value there. If in the future we allow providing a value from a variable, we can revisit to consider this case as you said!

**Pull request merged**
